### DISCUSSION TITLE
C# port of FLAML library

### DISCUSSION THREAD

**Discussion by torronen** [microsoft/FLAML#409] [General]
Are there plans to create a C# port of FLAML library? Or any community projects you know of doing it?

**Comment by torronen** [comment]
If not, would be feasible to run FLAML for LightGBM to get tuned parameters, and then train with said parameters inside C# application with Microsoft.ML.LightGBM library?

**Reply by sonichi** [reply]
FLAML has a C# implementation in [ML.NET Model Builder](https://dotnet.microsoft.com/apps/machinelearning-ai/ml-dotnet/model-builder) as part of [Visual Studio](https://visualstudio.microsoft.com/) 2022.
The way you mentioned should also work, as long as the LightGBM version matches.
@JakeRadMSFT or @LittleLittleCloud may have a better answer.

**Reply by torronen** [reply]
Thanks. I am aware of it being available through ML.NET Model Builder, but model builder does not yet allow to set a sampling key or custom validation data so I am not yet able to use it for my use case. Actually, my experience with Model Builder and ML.NET class library has been that model builder has much better tuner and that is why I am looking forward to using FLAML. LightGBM also has lots of powerful configurations which would not be available through Model Builder, such as setting for imbalanced classes.

So, I might try tuning with Python, and then run the production app in .NET with the parameters. I will pay special attention to LightGBM version.

**Reply by LittleLittleCloud** [reply]
@torronen 

We are about to start [Sweepable Api migration](#https://github.com/dotnet/machinelearning/pull/5993) pretty soon which will have the exact same search space && tuner used in model builder and flaml. I'll update under this thread when they are available publicly

**Reply by torronen** [reply]
That sounds great!

In principle it seems getting the parameters from FLAML to C# LightGBM seems to work, but I dont have any metrics yet. The names of parameters are slightly different but documentation is adequate to match them. Microsoft.ML seems to have version 2.3.1 of LightGBM.

@sonichi Another approach that might be useful, especially for anyone working with .NET, would be having some samples about conversion to ONNX. Maybe even some additional tooling.

**Reply by sonichi** [reply]
@torronen thanks for the feedback. I'll create an issue for that.

**Comment by torronen** [comment]
Hard time getting same results, C# has much lower accuracy. I will continue by comparing defaults of C# and Python.
LightGBM 2.31.

Currently, setting LightGBM variables as below, keeping others as default.

  ```
              LearningRate = learning_rate,
                MaximumBinCountPerFeature = max_bin,
                MinimumExampleCountPerLeaf = min_child_samples,
                NumberOfIterations = n_estimators,
                NumberOfLeaves = num_leaves,

                Booster = new GossBooster.Options
                {
                    L1Regularization = reg_alpha,
                    L2Regularization = reg_lambda,
                    FeatureFraction = colsample_bytree,
                },
```

**Reply by torronen** [reply]
After manually setting same default values as LightGBM 2.3.1 (below), I am still not able to reproduce performance reported by FLAML. Let me know if I should open an issue at LightGBM or Microsoft.ML. I am doing binary classification. 

```
// defaults https://github.com/microsoft/LightGBM/blob/v2.3.1/include/LightGBM/config.h
MinimumExampleCountPerGroup = 100, // min_data_per_group 
MaximumCategoricalSplitPointCount = 32, // max_cat_threshold
CategoricalSmoothing = 10.0, // cat_smooth 
NumberOfThreads = 0,
HandleMissingValue = true, //use_missing
UnbalancedSets = false, //is_unbalance
Sigmoid = 1.0, //sigmoid
L2CategoricalRegularization = 10, // cat_l2
// UseCategoricalSplit = could not find from Python settings
                
Booster = new GossBooster.Options
{
                    // defaults https://github.com/microsoft/LightGBM/blob/v2.3.1/include/LightGBM/config.h
                    SubsampleFraction = 1.0, // bagging_fraction  
                    SubsampleFrequency = 0, //bagging_freq 
                    TopRate = 0.2, // top_rate 
                    OtherRate = 0.1, // other_rate 
                     MaximumTreeDepth = 0, // max_depth. Python has -1, but Intellisense says 0 is for unlimited
                    // MinimumChildWeight = could not find from Python settings;
                    MinimumSplitGain = 0.0 // min_gain_to_split
                
```

If we are not able to find the cause, then I think next I will

- first, export as ONNX, retrain in python as needed, use in C#
- then, wait for FLAML C# API

**Reply by sonichi** [reply]
@LittleLittleCloud @JakeRadMSFT any suggestion?

**Reply by LittleLittleCloud** [reply]
What's the dataset you are using and what's the featurizer of your training pipeline.

**Reply by torronen** [reply]
I noted I may be using the wrong booster.   c++ [source ](https://github.com/microsoft/LightGBM/blob/v2.3.1/include/LightGBM/config.h )says gdbt is the default, but my code (sample from Microsoft Docs) uses Goss

**Reply by torronen** [reply]
@LittleLittleCloud This is a private client dataset, but I will do the testing next with some of the common ones. I will put the project on Github as well.

The featurizer is actually from template by Model Builder, slightly edited. It is a good point that it would be easy that Pandas and ML.NET pipelines would not match, but I think they do match:

For strings:
 - In ML.NET, for each separately: .`Append(mlContext.Transforms.Categorical.OneHotEncoding(....`
- In Python, make each category:   `train_set[textcols].apply(pd.Series.astype, dtype='category')`

For numbers:
 - In ML.NET only concatenate to features
  - In Python, downgrade them in pandas to lowest int or float

**Reply by torronen** [reply]
GossBooster.Options => GradientBooster.Options brings results closer, but still with some margin.

**Reply by torronen** [reply]
Maybe it is about the random seeds because FLAML uses col_subsamples? I have explicitly set them to match. I might play on weekend with them. If you happen to know: does the random seed in FLAML get passed to LightGBM and also does random seed of MLComtext get passed to LightGBM?

According to docs, LightGBM has defaults for different seeds but if random_seed is set then that seed is used to generate the others.

https://lightgbm.readthedocs.io/en/latest/Parameters.html

**Reply by sonichi** [reply]
> Maybe it is about the random seeds because FLAML uses col_subsamples? I have explicitly set them to match. I might play on weekend with them. If you happen to know: does the random seed in FLAML get passed to LightGBM and also does random seed of MLComtext get passed to LightGBM?
> 
> According to docs, LightGBM has defaults for different seeds but if random_seed is set then that seed is used to generate the others.
> 
> https://lightgbm.readthedocs.io/en/latest/Parameters.html

The random seed in FLAML is not passed to LightGBM. The final LightGBM model is something like LGBMRegressor(colsample_bytree=0.8025848209352517,
              learning_rate=0.09100963138990374, max_bin=255,
              min_child_samples=42, n_estimators=363, num_leaves=216,
              reg_alpha=0.001113000336715291, reg_lambda=76.50614276906414,
              verbose=-1)

One idea about the difference: flaml does some data preprocessing before hyperparameter search starts. You can access the transformed training data from `automl._X_train_all` if you set `keep_search_state=True` in `automl.fit()`. In some cases the transformed data are no different from the original data, and in other cases they are different.

**Reply by torronen** [reply]
@sonichi do you know if FLAML sets the seed for LightGBM or LightGBM should use default seeds? I could not yet find it from the source


in Microsoft.ML I believe seed is not set, except if I set it manually.

It seems the results vary wildly based on the seed. Getting closer by not setting the seed in c# but still ML.net is outperforming with same params

Another issue I might have is how to confirm the order of columns from pandas and Microsoft.ML transformer is exactly the same.

**Reply by torronen** [reply]
(for anyone who might be attempting the same, sigmoid seems to have different default in ML.net than c# source)

**Reply by sonichi** [reply]
> @sonichi do you know if FLAML sets the seed for LightGBM or LightGBM should use default seeds? I could not yet find it from the source
> 
> in Microsoft.ML I believe seed is not set, except if I set it manually.
> 
> It seems the results vary wildly based on the seed. Getting closer by not setting the seed in c# but still ML.net is outperforming with same params
> 
> Another issue I might have is how to confirm the order of columns from pandas and Microsoft.ML transformer is exactly the same.

LightGBM in flaml uses default seeds.

**Reply by torronen** [reply]
@sonichi @LittleLittleCloud 
I was mistaken, ML.NET sets a random seed if not set explicitly. Because FLAML lets LightGBM use default seeds and ML.NET does not allow setting the more specific seeds it causes problems:   https://github.com/dotnet/machinelearning/issues/6062

I believe this is the main cause for the difference. I may verify by using custom learner in FLAML and set random seed explicitly, or with custom ML.NET build.

**Reply by torronen** [reply]
After removing col_subsample / feature_fraction and setting random seed manually on both, the results are still not fully correlated. 
The best from today with 30 min time budget:
ML.NET Accuracy: 0.5419
Python accuracy= 0.5582
Train and validation sets are the same.

I dont know where to look next, but I will put some questions about observations on MachineLearning repo:
https://github.com/dotnet/machinelearning/issues/6063

**Reply by torronen** [reply]
( BTW, there was a long discussion about sigmoid, but I did not follow why sigmoid was not updated to LightGBM default https://github.com/dotnet/machinelearning/pull/667 )

**Reply by torronen** [reply]
@LitteLitteCloud here is simplified sample https://github.com/torronen/lightgbm-comparison

ML.NET pipeline 
https://github.com/torronen/lightgbm-comparison/blob/d63e9dc69b63b7b534284a19c13d19349c795029/mlnet/LightGBMTrainer.cs#L127-L135
Python pipeline (LightGBM version 2.3.1 should be used)
https://github.com/torronen/lightgbm-comparison/blob/d63e9dc69b63b7b534284a19c13d19349c795029/python/flaml-inbuilt-automl.py#L25-L30

FLAML accuracy 98+%
LightGBM in ML.NET accuracy 90%
ModelBuilder Binary:
  - FastTree 98+%,  
  - LightGBM 92%
  
There maybe small difference in pipeline so it may need verification. The results above, especially ModelBuilder's low scores for LightGBM, suggest there may be some misconfiguration in LightGBM .NET interface. I put some what I think here https://github.com/dotnet/machinelearning/pull/6064/files but results still do not match.

**Reply by LittleLittleCloud** [reply]
Hi @torronen

Thanks for the investigate. I take a look at the example you provided and my best guess is the difference between flaml and model builder on lgbm trainer's performance could be caused by the difference on featurizers used by flaml and model builder.

Flaml ignores string column if there's no duplicates in its values and passes catagorical column as it is to lgbm trainer, while model builder  uses text featurizer to encode string column to numeric vectors, and encodes catagorical column to numeric vector as well.

The difference on featurizers might put huge impact on final result. After I set all string column to catagorical in model builder, which is similar, but not identical with how flaml featurize its dataset, the train result goes up to ~97% in model builder, which is close to what flaml has.

![image](https://user-images.githubusercontent.com/16876986/151876317-2d162229-38d1-4d2a-a9b5-00cd5eac02b3.png)

**Reply by torronen** [reply]
Thanks. I will retry, but I think I also had [set all of them as classification](https://github.com/torronen/lightgbm-comparison/blob/d63e9dc69b63b7b534284a19c13d19349c795029/mlnet/titanic-binary.mbconfig): `"IsCategorical": true `in all string columns. I did not run even close to 10k iterations. 

 The python pipeline also [makes all string values a categories](https://github.com/torronen/lightgbm-comparison/blob/d63e9dc69b63b7b534284a19c13d19349c795029/python/flaml-inbuilt-automl.py#L30) before FLAML, no matter there are duplicate values or not.

The c# pipeline makes them [one hot encoded and is copied from Model Builder generated file](https://github.com/torronen/lightgbm-comparison/blob/d63e9dc69b63b7b534284a19c13d19349c795029/mlnet/LightGBMTrainer.cs#L130-L134). With the parameters from FLAML the C# console app does not get close. I think, with same parameters, LightGBM binary from .NET and Python should arrive at almost same result (although not exactly, as they mention on LightGBM docs, number of threads and the compiler used may affect)

So, I am wondering if Model Builder will actually find another local optimum, similar to FLAML finding another local optimum with early stopping. FLAML early stopping  stops in 6 seconds at 258 iterations. So, I am concerned if .NET is not using all the performance of LightGBM. Another reason for my concern is that I could not follow how num_iterations, learning_rate, min childs per group are set:
https://github.com/dotnet/machinelearning/issues/6063
https://github.com/dotnet/machinelearning/pull/6064 

Depending on how 0.97 is rounded, difference to 98.7% can be relatively big if the number of mistakes matter in some use cases (possibly 2x better if only mistakes matter). If it would be about small differences in featurizers, then I think it could be on area for "easy" way to improve the accuracy of model builder.

**Reply by LittleLittleCloud** [reply]
Hi @torronen 

It looks like C# lgbm can handle **Numeric** categorical feature automatically so it's not necessary to set a single-type column as **categorical** even if it is. And doing so might affect training performance in an opposite way since model builder uses one-hot encoder to process numeric categorical feature.

So in short, you should be able to reach approximatly the same performance using lgbm in model builder comparing with what flaml lgbm does, by
- set all string column to categorical if it has duplicate value, and ignores those which doesn't.
- set all numeric column's **is_categorical** to false and let lgbm decide whether treat that column as cat or not

After doing so, the training performance on my end raise to ~98.85%, within ~60 seconds

![image](https://user-images.githubusercontent.com/16876986/152054219-2e5896bf-b31b-4350-896d-1950384cbd44.png)

![image](https://user-images.githubusercontent.com/16876986/152054092-d695a83e-2005-4efc-9022-42e0afa7b902.png)

**Reply by LittleLittleCloud** [reply]
Here's the column property I use, you should be able to just copy and paste it to replace the `ColumnProperties` value in your .mbconfig
```json
"ColumnProperties": [
      {
        "ColumnName": "survived",
        "ColumnPurpose": "Label",
        "ColumnDataFormat": "Boolean",
        "IsCategorical": false,
        "Type": "Column",
        "Version": 1
      },
      {
        "ColumnName": "pclass",
        "ColumnPurpose": "Feature",
        "ColumnDataFormat": "Single",
        "IsCategorical": false,
        "Type": "Column",
        "Version": 1
      },
      {
        "ColumnName": "sex",
        "ColumnPurpose": "Feature",
        "ColumnDataFormat": "String",
        "IsCategorical": true,
        "Type": "Column",
        "Version": 1
      },
      {
        "ColumnName": "age",
        "ColumnPurpose": "Feature",
        "ColumnDataFormat": "Single",
        "IsCategorical": false,
        "Type": "Column",
        "Version": 1
      },
      {
        "ColumnName": "sibsp",
        "ColumnPurpose": "Feature",
        "ColumnDataFormat": "Single",
        "IsCategorical": false,
        "Type": "Column",
        "Version": 1
      },
      {
        "ColumnName": "parch",
        "ColumnPurpose": "Feature",
        "ColumnDataFormat": "Single",
        "IsCategorical": false,
        "Type": "Column",
        "Version": 1
      },
      {
        "ColumnName": "ticket",
        "ColumnPurpose": "Feature",
        "ColumnDataFormat": "String",
        "IsCategorical": true,
        "Type": "Column",
        "Version": 1
      },
      {
        "ColumnName": "fare",
        "ColumnPurpose": "Feature",
        "ColumnDataFormat": "Single",
        "IsCategorical": false,
        "Type": "Column",
        "Version": 1
      },
      {
        "ColumnName": "cabin",
        "ColumnPurpose": "Feature",
        "ColumnDataFormat": "String",
        "IsCategorical": true,
        "Type": "Column",
        "Version": 1
      },
      {
        "ColumnName": "embarked",
        "ColumnPurpose": "Feature",
        "ColumnDataFormat": "String",
        "IsCategorical": true,
        "Type": "Column",
        "Version": 1
      },
      {
        "ColumnName": "boat",
        "ColumnPurpose": "Feature",
        "ColumnDataFormat": "Single",
        "IsCategorical": false,
        "Type": "Column",
        "Version": 1
      },
      {
        "ColumnName": "body",
        "ColumnPurpose": "Feature",
        "ColumnDataFormat": "Single",
        "IsCategorical": false,
        "Type": "Column",
        "Version": 1
      },
      {
        "ColumnName": "home.dest",
        "ColumnPurpose": "Feature",
        "ColumnDataFormat": "String",
        "IsCategorical": true,
        "Type": "Column",
        "Version": 1
      },
      {
        "ColumnName": "surname",
        "ColumnPurpose": "Feature",
        "ColumnDataFormat": "String",
        "IsCategorical": true,
        "Type": "Column",
        "Version": 1
      },
      {
        "ColumnName": "title",
        "ColumnPurpose": "Feature",
        "ColumnDataFormat": "String",
        "IsCategorical": true,
        "Type": "Column",
        "Version": 1
      },
      {
        "ColumnName": "firstname",
        "ColumnPurpose": "Feature",
        "ColumnDataFormat": "String",
        "IsCategorical": true,
        "Type": "Column",
        "Version": 1
      },
      {
        "ColumnName": "random",
        "ColumnPurpose": "Feature",
        "ColumnDataFormat": "Single",
        "IsCategorical": false,
        "Type": "Column",
        "Version": 1
      }
    ]
```

**Reply by torronen** [reply]
Thanks. I will try this and compare the parameters it give. Hopefully it will help solve the mystery and what might be the difference between the hyperparameters... BTW, I just realize accuracy will be different because Model builder does not use same test data set as python script or the .NET console app.

**Reply by torronen** [reply]
@LittleLittleCloud column boat is acutally text. I t has value like "5 9" and A-C. However, if according to your sample, it is put as Single then it gives better results. It would make sense because "boat" indicates the lifeboat, as I understand. If a person has no boat specified, then chance of survival is almost 0%. 

But why string wouldn't do? If we set as numeric, then maybe we loose some performance because A-C can not be parsed as numbers?

**Reply by torronen** [reply]
Boat as categorical string drops to 95%, but most iterations are at 91%

**Reply by LittleLittleCloud** [reply]
My best guess is if column boat is put as cat `String`, it will be processed by one-hot encoder which increases dimension of features that put into lgbm and might end up with a lower validate result because dataset with higher dimension is more prone to overfitting (curse of dimension). That's also probably why after removing `IsCatagorical` for all numeric features, the validation result increase. Because that change helps decrease dimensions.

I would verify it via using hash encoding instead. If I'm right we should see the training result reachs to 99% in a faster speed.

**Comment by torronen** [comment]
@sonichi If random seed makes a big difference in some datasets it might have some implications for the optimal search space. Choice of booster had also quite big difference. I might try to "tune" seed and booster. Would you or your team have suggestions on best way to do it? 

- Just add to the list of parameters to tune?
- first tune without any randomity based feature, then final tune with randomity and seed?
- full tuning in a loop with different seeds?
Probably best to "tune" each specific seed separately, not just random_seed

I have one concern about it: If the tuning results rely on randomly dropped columns and it matters a lot which columns gets dropped, then if the next dataset version removes a column from the beginning of file the tuning results are no longer valid and tuning needs to be redone. 

Not every developer might realize a small change in dataset, like remove column or add a engineer a new column in pipeline, migth break the optimal values. Therefore, maybe, for default search space it should be considered not to use randomity based ones at all? Microsoft.ML would not seem to use it. Tuned results from Microsoft.ML are fairly close to FLAML but the tree is much bigger in Microsoft.ML.

For experimentation, we may need a public dataset with high number of columns (+high risk of overfit) so that colsample_bytree becomes effective. I can not share the current dataset, but colsample_bytree values has lowest 0.67 and highest 0.9 (depending on metric) and it has 2500 columns.

**Reply by sonichi** [reply]
> @sonichi If random seed makes a big difference in some datasets it might have some implications for the optimal search space. Choice of booster had also quite big difference. I might try to "tune" seed and booster. Would you or your team have suggestions on best way to do it?
> 
> * Just add to the list of parameters to tune?
> * first tune without any randomity based feature, then final tune with randomity and seed?
> * full tuning in a loop with different seeds?
>   Probably best to "tune" each specific seed separately, not just random_seed
> 
> I have one concern about it: If the tuning results rely on randomly dropped columns and it matters a lot which columns gets dropped, then if the next dataset version removes a column from the beginning of file the tuning results are no longer valid and tuning needs to be redone.
> 
> Not every developer might realize a small change in dataset, like remove column or add a engineer a new column in pipeline, migth break the optimal values. Therefore, maybe, for default search space it should be considered not to use randomity based ones at all? Microsoft.ML would not seem to use it. Tuned results from Microsoft.ML are fairly close to FLAML but the tree is much bigger in Microsoft.ML.
> 
> For experimentation, we may need a public dataset with high number of columns (+high risk of overfit) so that colsample_bytree becomes effective. I can not share the current dataset, but colsample_bytree values has lowest 0.67 and highest 0.9 (depending on metric) and it has 2500 columns.

For tuning, just add to the search space of the inherited custom estimator. For not tuning, remove the hp from search space of the inherited custom estimator. How much a small change of the dataset affects the result depends on your specific use case. BTW, what application do you use flaml for?

**Reply by torronen** [reply]
Thanks. 

For now I am just testing. This specific dataset is for e-commerce click flow to predict if user will continue session or not. For now, lots of low or no correlation features, few probably have more correlation (session length, pages loaded... screen size, mouse position, scroll position... http headers...) I am planning to use with AI-enabled school or gov (applicaton form processing, job recomendations) or B2B tools (niche focus, like the click flow or chatbot improvement). Datasets often have high number of features, as some other persons in our team are tasked with data sourcing and they add everything they can find, my task is to drop what is not needed.

**Reply by sonichi** [reply]
Thanks for sharing. One useful feature might be [warm start](https://microsoft.github.io/FLAML/docs/Use-Cases/Task-Oriented-AutoML#warm-start), when your data are changed slightly.
